rero_ils/modules/serializers/base.py

```diff
@@ -0,0 +1,192 @@
+
+"""RERO ILS record serialization."""
+import inspect
+from abc import ABC
+
+from flask import json, request, url_for
+from invenio_records_rest.serializers.json import \
+    JSONSerializer as _JSONSerializer
+
+from rero_ils.modules.documents.utils import filter_document_type_buckets
+
+
+class PostprocessorMixinInterface(ABC):
+    """Mixin postprocessing records during serialization."""
+
+    def postprocess_serialize_search(self, results, pid_fetcher):
+        """Post-process the search results.
+
+        :param results: Search result.
+        :param pid_fetcher: Persistent identifier fetcher.
+        """
+        raise NotImplementedError()
+
+
+class PostprocessorMixin(PostprocessorMixinInterface):
+    """Base class for post-processing record serializers."""
+
+    def postprocess_serialize_search(self, results, pid_fetcher):
+        """Post-process the search results.
+
+        Post-processing a search result is a three steps process :
+          1) allow modification on each search hit
+          2) allow modification on aggregation buckets
+          3) allow modification about links part
+
+        :param results: Search result.
+        :param pid_fetcher: Persistent identifier fetcher.
+        """
+        for hit in results.get('hits', {}).get('hits', []):
+            self._postprocess_search_hit(hit)
+        if aggregations := results.get('aggregations'):
+            self._postprocess_search_aggregations(aggregations)
+        self._postprocess_search_links(results, pid_fetcher)
+        return results
+
+    def _postprocess_search_links(self, search_results, pid_fetcher) -> None:
+        """Post-process search links.
+
+        :param search_results: Elasticsearch search result.
+        :param pid_fetcher: Persistent identifier fetcher related to records
+                            into the search result.
+        """
+        # add REST API to create a record related to the search result.
+        pid_type = pid_fetcher(None, {'pid': '1'}).pid_type
+        url = url_for(f'invenio_records_rest.{pid_type}_list', _external=True)
+        search_results['links'].update({'create': url})
+
+    def _postprocess_search_hit(self, hit: dict) -> None:
+        """Post-process a specific search hit.
+
+        :param hit: the dictionary representing an ElasticSearch search hit.
+        """
+        # DEV NOTES :
+        #   Override this method in subclass to operate specific
+        #   modification/enrichment on a search hit.
+
+    def _postprocess_search_aggregations(self, aggregations: dict) -> None:
+        """Post-process aggregations from a search result.
+
+        :param aggregations: the dictionary representing ElasticSearch
+                             aggregations section.
+        """
+        if 'document_type' in aggregations:
+            aggr = aggregations['document_type'].get('buckets')
+            filter_document_type_buckets(aggr)
```

Comment:
because, in any cases (for RERO-ILS serializing), if the search contains a `document_type` facets, this facets must be cleaned. I think it was often use with Virtua loaded records ; Now it's impossible to enter an invalid subtype against the JSONSchema validation.